enhancements/cluster-logging/elasticsearch-cert-management.md

```diff
@@ -0,0 +1,158 @@
+---
+title: elasticsearch-cert-management
+authors:
+  - "@ewolinetz"
+reviewers:
+  - "@jcantrill"
+  - "@bparees"
+  - "@alanconway"
+  - "@jpkrohling"
+approvers:
+  - "@jcantrill"
+  - "@bparees"
+  - "@alanconway"
+  - "@jpkrohling"
+creation-date: 2021-01-05
+last-updated: 2021-01-05
+status: implementable
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# elasticsearch-cert-management
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Migration plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently to use the Elasticsearch Operator, one would need to generate secrets in a specific manner which creates a high amount of complexity for use. This proposal seeks to outline a mechanism where the Elasticsearch Operator creates and maintains these certificates and the elasticsearch/kibana secret instead of other operators (e.g. Cluster Logging and Jaeger), and allows an annotation mechanism on secrets for injecting required keys and certificates for mTLS with Elasticsearch.
+
+## Motivation
+
+Currently our managed Elasticsearch cluster requires anything using our operator to generate, manage, and provide certs and secrets for the operator and its cluster to use. This leads to many potential sources to create certificates and can cause a gap between what the managed Elasticsearch requires and what something may provide. This would instead make the Elasticsearch Operator a single source of truth for how certificates would be generated.
+
+### Goals
+The specific goals of this proposal are:
+
+* Outline the responsibility for the Elasticsearch Operator regarding certificates
+
+* Expose the desired annotation that the operator would watch for on secrets where an elasticsearch CR exists
+
+* Discuss upgrade strategies from the current LTS mechanism.
+
+We will be successful when:
+
+* Users of Elasticsearch Operator no longer need to maintain certificates for Elasticsearch and can have certificates injected into their secrets
+
+* There is a clear path forward for automatic upgrades
+
+### Non-Goals
+
+* This does not seek to implement work to adhere to the cluster level TLS definitions, however this does make the implementation less complex for the Elasticsearch use-case.
+
+## Proposal
+
+Instead of the current implementation where any consumer of the Elasticsearch Operator is responsible for providing certificates and secrets for Elasticsearch and the operator to use, we want to shift this responsibility to the Elasticsearch Operator instead. This will greatly improve the use case of the operator and make the operator the single source of truth for its certificates.
```

Comment:
I don't think this is going to be a breaking change, but I don't remember how it's all tied together from the top of my head. Could you state explicitly in the doc that we will still be able to use the current mechanism, as we have two release streams that we'd not change to use this new feature?

Reply:
currently Jaeger Operator runs a script that will create the certs and secrets that are provided for the elasticsearch cluster that is to be used. Off the top of my head I'm not sure how we could support the currently used model and this proposed one at the same time.

I would suggest that the Jaeger Operator not use an updated EO version, but given it is a dependency of (at least) two operators that isn't a likely solution.

Reply:
We'll need a good coordination then before this feature gets released, and include the Maistra team (cc @knrc and @rcernich) in the loop.

Reply:
Thanks for the head's up @jpkrohling.  Maistra has no direct integration with elasticsearch.  Anything we might configure is done through the jaeger cr.

Reply:
Is there not a path where by which we can support both where one takes precedence over the other.  For example, if you tell me the secret, then I will use it unless it has say an annotation on it, then the EO injects the files into the secret?  I believe the convention now is we just expect the secret to exist and be named the same as the `elasticsearch` resource

Reply:
how would we know who created the secret though? we would see that there is a secret out there which is the expected outcome no matter who is managing certs...

Reply:
Couldnt  we assume for a smoother transition:
* follow existing pattern of a named secret that matches named elasticsearch resource 
or
* elasticsearch resource with annotation takes precedence over previous pattern?

I think we want the requestor (CL/Jaeger) to own the secret regardless as it has to exist to be mounted into something like the collector pod, and assume the secret will be injected the explicit keys.  I defer to @alanconway here

Reply:
this enhancement outlines that to have EO create and manage certs for your cluster, you'll need to annotate your elasticsearch CR. If you are doing that, then you are opting in to this process and it would replace the current pattern (where CL and Jaeger manage certs/secrets).

> I think we want the requestor (CL/Jaeger) to own the secret regardless as it has to exist to be mounted into something like the collector pod, and assume the secret will be injected the explicit keys. I defer to @alanconway here

I think we can do this either way... I would rather lean towards doing what the cert signing service already does though to be consistent.